### ISSUE TITLE
C build warnings in scipy.special under python 3.5

### ISSUE THREAD

**Issue by ev-br** [scipy/scipy#6144]
Building scipy under python 3.5 generates, among other things, a bunch of warnings of this sort:

```
scipy/special/_ufuncs.c: In function ‘__pyx_f_5scipy_7special_11_ellip_harm_lame_coefficients’:
scipy/special/_ufuncs.c:25858:3: warning: ‘__pyx_v_t’ may be used uninitialised in this function [-Wuninitialized]
scipy/special/_ufuncs.c: In function ‘__pyx_f_5scipy_7special_11_ellip_harm_ellip_harmonic’:
scipy/special/_ufuncs.c:26646:14: warning: ‘__pyx_v_psi’ may be used uninitialised in this function [-Wuninitialized]
scipy/special/_ufuncs.c:26448:10: note: ‘__pyx_v_psi’ was declared here
scipy/special/_ufuncs.c:26626:18: warning: ‘__pyx_v_size’ may be used uninitialised in this function [-Wuninitialized]
scipy/special/_ufuncs.c:26442:7: note: ‘__pyx_v_size’ was declared here
scipy/special/_ufuncs.c: In function ‘__pyx_f_5scipy_7special_7_cunity_cexpm1’:
scipy/special/_ufuncs.c:21079:31: warning: ‘__pyx_v_ezr’ may be used uninitialised in this function [-Wuninitialized]
scipy/special/_ufuncs.c: In function ‘__pyx_f_5scipy_7special_7_legacy_ellip_harmonic_unsafe’:
scipy/special/_ufuncs.c:26646:14: warning: ‘__pyx_v_psi’ may be used uninitialised in this function [-Wuninitialized]
scipy/special/_ufuncs.c:26448:10: note: ‘__pyx_v_psi’ was declared here
scipy/special/_ufuncs.c:26626:18: warning: ‘__pyx_v_size’ may be used uninitialised in this function [-Wuninitialized]
scipy/special/_ufuncs.c:26442:7: note: ‘__pyx_v_size’ was declared here
scipy/special/_ufuncs.c: In function ‘__pyx_f_5scipy_7special_17_spherical_bessel_spherical_jn_d_real’:
scipy/special/_ufuncs.c:35420:126: warning: ‘__pyx_v_sn’ may be used uninitialised in this function [-Wuninitialized]
scipy/special/_ufuncs.c: In function ‘__pyx_f_5scipy_7special_17_spherical_bessel_spherical_yn_d_real’:
scipy/special/_ufuncs.c:35568:126: warning: ‘__pyx_v_sn’ may be used uninitialised in this function [-Wuninitialized]
```

These should probably be checked whether these are just noise from Cython, or is there something more to it.

**Comment by person142** [comment]
I checked; they are all noise. The ones related to ``ellip_harm.pxd`` come from blocks like this:
```python
if p - 1 < r + 1:
    t, tp, size = 'K', p, r + 1
elif p - 1 < (n - r) + (r + 1):
    t, tp, size = 'L', p - (r + 1), n - r
elif p - 1 < (n - r) + (n - r) + (r + 1):
    t, tp, size = 'M', p - (n - r) - (r + 1), n - r
elif p - 1 < 2*n + 1:
    t, tp, size = 'N', p - (n - r) - (n - r) - (r + 1), r
```
The compiler doesn't like ending with the ``elif``. The validity of ``p``, however, is checked elsewhere (and the function bails if it doesn't satisfy one of these four cases), so it's not a problem. (This block appears twice, and when it appears in ``lame_coefficients`` you could move the validity checks into the block and silence a couple of the warnings. You could add redundant checks the other time it appears in ``ellip_harm_eval`` and silence all of them.)

The ones from ``_cunity.pxd`` come from this block:
```python
if zr <= -40:
    x = -1.0
else:
    ezr = expm1(zr)
    x = ezr*cos(zi) + cosm1(zi)
# don't compute exp(zr) too, unless necessary
if zr > -1.0:
    y = (ezr + 1.0)*sin(zi)
else:
    y = exp(zr)*sin(zi)
```
The compiler is warning about ``ezr``. The block is hard to parse but correct, and I don't see a clearer way to write it.

The ones from ``_spherical_bessel.pxd`` come from blocks like this:
```python
if n < 0:
    sf_error.error("spherical_jn", sf_error.DOMAIN, NULL)
    return nan

...

s0 = sin(x)/x
if n == 0:
    return s0
s1 = (s0 - cos(x))/x
if n == 1:
    return s1

for idx in range(n - 1):
    sn = (2*idx + 3)*s1/x - s0
    s0 = s1
    s1 = sn
    if npy_isinf(sn):
        # Overflow occurred already: terminate recurrence.
        return sn

return sn
```
It doesn't like ``sn``. This code is clear and any restructuring I can think of would make it less clear.

**Comment by rgommers** [comment]
checked, so closing - we have lots of build warnings left, so no need to keep open issues for specific ones. also, looks like @person142 fixed a bunch of these
